### PR TITLE
fix mysql deployment strategy

### PR DESCRIPTION
**Pull request by thesuperzapper**
When trying to restart MySQL instances, they get stuck because the old container will refuse to give up the lock on the DB:
* `kubectl rollout restart deployment/katib-mysql  -n kubeflow`
* `kubectl rollout restart deployment/metadata-db  -n kubeflow`

This PR fixes that by setting the deployment strategy to `Recreate`.

**Checklist:**
- [X] Unit tests have been rebuilt: 
    1. `cd manifests/tests`
    2. `make generate-changed-only`
    3. `make test`

### CODE REVIEW

**Comment by thesuperzapper** [comment]
/assign @andreyvelich @zhenghuiwang

can you please review?

**Comment by andreyvelich** [comment]
@thesuperzapper Thank you for updating this, have you checked that when `katib-mysql` pod is restarted `katib-db-manager` connects to the new DB properly ?

**Comment by thesuperzapper** [comment]
@andreyvelich well as far as I can see it works

**Comment by andreyvelich** [comment]
@thesuperzapper Thank!
Can you open a PR to make changes in Katib [manifests](https://github.com/kubeflow/katib/blob/master/manifests/v1beta1/mysql-db/deployment.yaml) also, please?

**Review by andreyvelich — COMMENTED** [review]
/lgtm

**Comment by thesuperzapper** [comment]
Gentle ping: @gaocegege @johnugeorge

**Comment by thesuperzapper** [comment]
Have you had a chance to review this? @andreyvelich @zhenghuiwang

**Review by andreyvelich — APPROVED** [review]
@thesuperzapper Sorry for the long reply!
/lgtm
/approve

**Comment by thesuperzapper** [comment]
I still need one of the metadata owners to LGTM this

Can one of @neuromage @prodonjs @zhenghuiwang please take a look?

**Comment by PatrickXYS** [comment]
Can you try re-run new test?

/test ?

**Comment by thesuperzapper** [comment]
I am still waiting on one of @neuromage  @prodonjs  @zhenghuiwang to LGTM this

I want to get this in before Kubeflow 1.2

**Comment by PatrickXYS** [comment]
@thesuperzapper I don't think any of them are active in kubeflow community now.

You might need to find review from other one, after that, try re-run test because we have migrated to shared test-infra

File: metadata/overlays/db/metadata-db-deployment.yaml
Comment (Jeffwan) [inline metadata/overlays/db/metadata-db-deployment.yaml:12]: em. Technically, we should use `statefulset` for mysql. because pod under deployment is controlled by ReplicaSet and replacement will be created immediately if you do an upgrade.  

This workaround has some down side, existing pods will be killed before new ones created. The problem is these are not all DB layers. `Recreate` strategy may cause downtime to its dependencies like KFP or Katib components. 
It requires this layer has the resilience on db failure. 

I have not tested this and not sure if anything abnoarmal there.

Comment (thesuperzapper) [reply metadata/overlays/db/metadata-db-deployment.yaml:12]: @Jeffwan if you see this PR's description, currently, if you try and restart the MySQL Deployment, it will hang forever.

This is because the old Pod wont give up the lock on the DB files, so the new one can never be "ready" unless you manually kill the old Pod.

Comment (Jeffwan) [reply metadata/overlays/db/metadata-db-deployment.yaml:12]: I saw the problem.. My asking is if you see abnormal on katib or KFP side when you did upgrade with `Recreate ` strategy because DB service will have down time. connection maybe lost and if application doesn't handle this well, it could be an issue..

Comment (thesuperzapper) [reply metadata/overlays/db/metadata-db-deployment.yaml:12]: @Jeffwan, we don't really have a choice, there will have to be some downtime as only one MySQL instance can lock the DB files at a time.

We just need to design Katib / KFP to be able to handle short downtimes of MySQL (like using exponential-backoff retries)

Comment (Jeffwan) [reply metadata/overlays/db/metadata-db-deployment.yaml:12]: sounds good

**Comment by andreyvelich** [comment]
@thesuperzapper Can you try to rebase, to fix the tests?

**Comment by thesuperzapper** [comment]
@andreyvelich @Jeffwan I have rebased

**Comment by andreyvelich** [comment]
@thesuperzapper I think you need to run `make generate-changed-only` again to fix unit tests.

**Comment by thesuperzapper** [comment]
@andreyvelich when I run `make generate-changed-only` on a rebased version of my branch, I get an error, is master broken? Or is it something to do with my setup?

**Comment by PatrickXYS** [comment]
@thesuperzapper If there's something wrong in unit test, you might need to go figure it out, keeping unit test succeed is the bottom line.

**Comment by Jeffwan** [comment]
@thesuperzapper Can you share the errors? What's the errors msg? What kustomize version you have in your environment. Recent change all pass UT and I think master should be good.

**Comment by thesuperzapper** [comment]
@Jeffwan it seems like kustomize 3.8.1 is not working with our repo, we probably need to fix this for users

I have rolled my local version back to 3.6.1 and `make generate-changed-only` works again

**Comment by PatrickXYS** [comment]
Copy from manifest README

```
This repo contains kustomize packages for deploying Kubeflow applications.

If you are a contributor authoring or editing the packages please see Best Practices. Note, please use kustomize v3.2.1 with manifests in this repo, before #538 is fixed which will allow latest kustomize to be used.
```

https://github.com/kubeflow/manifests

**Comment by thesuperzapper** [comment]
@PatrickXYS, yeah, but we should aim to support any widely used versions of kustomize

**Comment by PatrickXYS** [comment]
@thesuperzapper Ideally we should do so, let's create a GitHub issue to talk about it.

/lgtm
/approve
/hold

Can you cherry-pick this PR into v1.2-branch as well?

**Comment by Jeffwan** [comment]
this should be good to go.

/hold cancel

**Comment by thesuperzapper** [comment]
@PatrickXYS, I dont think I have the required permissions to chery-pick this commit

**Comment by PatrickXYS** [comment]
@thesuperzapper You won't be able to write directly to v1.2-branch, but send a cherry-pick PR

**Pull request merged**
